### PR TITLE
drivers/w5100: Fixed netdev_driver_t::recv() API

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

The current implementation of the `netdev_driver_t::recv()` API in the W5100 driver does not comply with documentation:

 - The drop feature of `recv()` is not implemented
 - In case of a too small buffer no error is returned and the frame is silently truncated. The API mandates returning `-ENOBUFS`

### Testing procedure

It would be nice to add a test application for this, as many drivers are affected...

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/issues/10410, https://github.com/RIOT-OS/RIOT/issues/10413

----
Update 1: Added reference to #10413

### CODE REVIEW

**Comment by maribu** [comment]
Please note that I do not have the hardware to test. This PR is thus completely *untested*.

**Comment by maribu** [comment]
I just updated the second commit to comply with https://github.com/RIOT-OS/RIOT/issues/10413

**Comment by aabadie** [comment]
This driver was initially written by @haukepetersen, so maybe he can help testing this PR or maybe @miri64 also has access to this hardware ?

File: drivers/w5100/w5100.c [hunk 1]
```diff
@@ -247,21 +248,29 @@ static int recv(netdev_t *netdev, void *buf, size_t len, void *info)
...
 
-        DEBUG("[w5100] recv: got packet of %i byte (at 0x%04x)\n", n, (int)rp);
+        DEBUG("[w5100] recv: got packet of %i byte (at 0x%04x)\n", len, (int)rp);
 
         /* read the actual data into the given buffer if wanted */
         if (in_buf != NULL) {
+            /* Is provided buffer big enough? */
+            if (len > max_len) {
+                spi_release(dev->p.spi);
+                return -ENOBUFS;
```
Comment (jia200x) [inline drivers/w5100/w5100.c]: From docs:
```
If called with buf != NULL and len is smaller than the received packet:
    The received packet is dropped
    The content in buf becomes invalid. (The driver may use the memory to implement the dropping - or may not change it.)
    -ENOBUFS is returned
```

AFAIS, the packet won't be dropped if `len > max_len`. The frame is not removed from the RX buffer

Comment (maribu) [reply drivers/w5100/w5100.c]: Yes, sorry. I created this PR when the exact semantics of that case where still in discussion and I forgot to update this PR to the final decision. (I will check tomorrow the other related PRs as well, as they might have similar issues.)

**Comment by jia200x** [comment]
In 079ea831d98d49231577a1fd8064f5e7a2bbc62d commit message:

> the frame should not be droppend and
no data of the frame should be returned.

1. s/droppend/dropped
2. It should say "the frame **should** be dropped", as stated in the documentation

**Comment by jia200x** [comment]
I don't have hardware to test neither :/

**Comment by maribu** [comment]
@jia200x: Squashed and fixed the commit message as requested.

**Comment by jia200x** [comment]
@miri64 or @jcarrano, could you give it a try with hardware at FU?

File: drivers/w5100/w5100.c [hunk 2]
```diff
@@ -247,36 +260,38 @@ static int recv(netdev_t *netdev, void *buf, size_t len, void *info)
...
-            wreg(dev, S0_CR, CR_RECV);
+                  len, (int)rp);
+        }
 
-            /* if RX buffer now empty, clear RECV interrupt flag */
-            if ((num - psize) == 0) {
-                wreg(dev, S0_IR, IR_RECV);
-            }
+        /* if frame received OR drop requested, remove frame from RX buffer */
+        if ((in_buf != NULL) || (len > 0)) {
```
Comment (miri64) [inline drivers/w5100/w5100.c]: I think this must be `&&`?

Comment (miri64) [reply drivers/w5100/w5100.c]: (when I applied this it worked)

Comment (maribu) [reply drivers/w5100/w5100.c]: This is strange. The original code did advance the read pointer address after reading the frame, which is now done in `drop()`. When `||` is changed to `&&` above, the read pointer address would no longer be advanced after reading the frame. I will try to investigate this issue

https://github.com/RIOT-OS/RIOT/blob/af627606f08fd9db74b7f8b927e6b84f43fbe562/drivers/w5100/w5100.c#L254-L273

Comment (maribu) [reply drivers/w5100/w5100.c]: OK, it should have checked for `max_len > 0`, as `len` is computed within this function and always `>0`, unless an empty frame was received.

Also, checking for `max_len > 0` should be sufficient. The cases we have are:

| Case     | `buf`     | `max_len` | Call `drop()`? |
|----------|-----------|-----------|----------------|
| get size | `NULL`    | `0`       | No             |
| recvieve | `!= NULL` | `> 0`     | Yes            |
| drop     | `NULL`    | `> 0`     | Yes            |

**Comment by maribu** [comment]
@miri64: Mind to check again?

**Comment by miri64** [comment]
I think something is broken still. While I sometimes am able to ping the laptop from the node I very quickly get:

```
2019-01-22 13:35:42,788 - INFO #  ping6 fe80::bbf2:ce3b:1a7:f6c8
2019-01-22 13:35:43,796 - INFO # ping timeout
2019-01-22 13:35:43,804 - INFO # dropping additional response packet (probably caused by duplicates)
2019-01-22 13:35:44,812 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,816 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,820 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,820 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,824 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,829 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,833 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,836 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,840 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,841 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,845 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,849 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,853 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,857 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,857 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,861 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,865 - INFO # gnrc_netif: possibly lost interrupt.
2019-01-22 13:35:44,869 - INFO # gnrc_netif: possibly lost interrupt.
```

And that doesn't stop. Mind you, that my setup seems to be a little flaky. Sometimes when I reset the board isn't able to establish the SPI connection and just crashes. :-/ With `ping -s 65535 <ADDR>` I unsurprisingly get a `EMSGSIZE` back from the `ping` application. `sudo ping -f <addr>` works until I get the `possibly lost interrupt` message.

**Comment by maribu** [comment]
OK, I was able to get an Arduino Ethernet Shield. I should be able to debug this PR

**Comment by maribu** [comment]
With the Arduino-Mega2560 this works for me as expected. (I was only able to use `examples/gnrc_minimal` due to the RAM limitations of the board.)

My test was:

```
$ ping6 -I eth0 -c 20 fe80::2223:23ff:fe23:2323; \
> ping6 -I eth0 -c 20 -s 65335 fe80::2223:23ff:fe23:2323; \
> ping6 -I eth0 -c 20 fe80::2223:23ff:fe23:2323
```

Output (without the `64 bytes from ...` lines) with this PR:

```
PING fe80::2223:23ff:fe23:2323(fe80::2223:23ff:fe23:2323) from :: eth0: 56 data bytes
--- fe80::2223:23ff:fe23:2323 ping statistics ---
20 packets transmitted, 20 received, 0% packet loss, time 52ms
rtt min/avg/max/mdev = 23.804/25.178/43.095/4.113 ms

PING fe80::2223:23ff:fe23:2323(fe80::2223:23ff:fe23:2323) from :: eth0: 65335 data bytes
--- fe80::2223:23ff:fe23:2323 ping statistics ---
20 packets transmitted, 0 received, 100% packet loss, time 264ms

PING fe80::2223:23ff:fe23:2323(fe80::2223:23ff:fe23:2323) from :: eth0: 56 data bytes
--- fe80::2223:23ff:fe23:2323 ping statistics ---
20 packets transmitted, 20 received, 0% packet loss, time 56ms
rtt min/avg/max/mdev = 23.811/24.980/43.101/4.161 ms
```

Output with current master (again eliding `64 bytes from ...`):
```
PING fe80::2223:23ff:fe23:2323(fe80::2223:23ff:fe23:2323) from :: eth0: 56 data bytes
--- fe80::2223:23ff:fe23:2323 ping statistics ---
20 packets transmitted, 20 received, 0% packet loss, time 50ms
rtt min/avg/max/mdev = 21.708/23.070/40.755/4.061 ms

PING fe80::2223:23ff:fe23:2323(fe80::2223:23ff:fe23:2323) from :: eth0: 65335 data bytes
--- fe80::2223:23ff:fe23:2323 ping statistics ---
20 packets transmitted, 0 received, 100% packet loss, time 276ms

PING fe80::2223:23ff:fe23:2323(fe80::2223:23ff:fe23:2323) from :: eth0: 56 data bytes
--- fe80::2223:23ff:fe23:2323 ping statistics ---
20 packets transmitted, 0 received, 100% packet loss, time 276ms
```

**Comment by maribu** [comment]
@miri64: Mind to test it with a reduced SPI clock, e.g. 1 MHZ instead of 5?

**Comment by miri64** [comment]
> @miri64: Mind to test it with a reduced SPI clock, e.g. 1 MHZ instead of 5?

Doesn't help :(. I retested this PR on master. There it is pretty stable.

BTW when I run into the "possibly lost interrupt" problem and then hit the reset button on the board, initialization fails as reported above (and only then). I only am able to get the device back into a usable state by disconnecting it from USB and replugging.

**Comment by maribu** [comment]
OK, previous code did update `rp` (without using the result) in the drop case in a very sneaky way (see fixup above). As the result of this update was never used, moving the code into `static inline void drop()` cannot have introduced a bug. But I still dropped the sneaky `rp += psize` as argument in a function call, as this will only confuse the reader.

**Comment by miri64** [comment]
Still having the `gnrc_netif: possibly lost interrupt` which I don't observe on master (there and in #10844 the device just hangs silently at some point).

**Comment by maribu** [comment]
@miri64: I bought an Arduino Due and I'm able to send 100 pings from Linux to the Arudino while simultaneously sending 100 pings from the Arduino to the Linux machine and got 98 response in each direction with this PR. I did so both using my "self-cooked" toolchain and `BUILD_IN_DOCKER=1`.

So I guess it works for me as expected, or did you test for longer periods of time?

If I'm indeed unable to reproduce your issue, I think chances are high that your board or shield is flanky. Or maybe there is some timing issue in the driver that by luck is not triggered by the code generated by the toolchains I used. Would you mind testing again with binaries generated via the docker build to rule out that the issue is triggered by your toolchain?

Btw: The W5100 draws on average 100mA [according to this site](https://www.itead.cc/wiki/W5100_Ethernet_Shield_V1.0) and the [Arudino Due](https://store.arduino.cc/arduino-due) is drawing up to 130mA. Some Notebooks do not comply with the USB standard when it comes to the current they provide. Have you tried to connect the Arudino on a Desktop PC or using a powered USB hub?

**Comment by miri64** [comment]
> If I'm indeed unable to reproduce your issue, I think chances are high that your board or shield is flanky. Or maybe there is some timing issue in the driver that by luck is not triggered by the code generated by the toolchains I used. Would you mind testing again with binaries generated via the docker build to rule out that the issue is triggered by your toolchain?

Might be. I'll try later to test with Docker again.

> Btw: The W5100 draws on average 100mA [according to this site](https://www.itead.cc/wiki/W5100_Ethernet_Shield_V1.0) and the [Arudino Due](https://store.arduino.cc/arduino-due) is drawing up to 130mA. Some Notebooks do not comply with the USB standard when it comes to the current they provide. Have you tried to connect the Arudino on a Desktop PC or using a powered USB hub?

Up until now all tests were done via a USB hub ;-).

**Comment by miri64** [comment]
> Might be. I'll try later to test with Docker again.

Very likely now. I don't get the device to initialize anymore. :( I guess then I just have to trust your testing. At least it was not worse than before (since we now at least have output during the endless loop). Please squash.

**Comment by maribu** [comment]
@jia200x: I think I addressed your remarks. Can you take a look and confirm?

**Comment by maribu** [comment]
@jia200x: Your "changes requested" is currently blocking this PR. I think I addressed them. Can you have a look?

**Comment by jia200x** [comment]
> @jia200x: Your "changes requested" is currently blocking this PR. I think I addressed them. Can you have a look?

Oh sorry. Sure, I will

**Comment by aabadie** [comment]
It's a bug fix, should be backport in 2019.01 ?

**Comment by miri64** [comment]
Yepp. @maribu can you provide one please?

**Pull request merged**